src/datajudge/db_access.py

```diff
@@ -782,6 +782,34 @@ def get_numeric_gaps(
     )
 
 
+def get_functional_dependency_violations(
+    engine: sa.engine.Engine,
+    ref: DataReference,
+    key_columns: list[str],
+):
+    selection = ref.get_selection(engine)
+    uniques = selection.distinct().cte()
+
+    key_columns_sa = [uniques.c[key_column] for key_column in key_columns]
+    violations_stmt = (
+        sa.select(*key_columns_sa).group_by(*key_columns_sa).having(sa.func.count() > 1)
+    ).cte()
+
+    join_condition = sa.and_(
```

Comment:
I fully side with you that the actual violations are very useful but I thought that `violation_stmt` is already derived from `unique` and that the join is thereby redundant.

Reply:
For example, suppose there were two columns A and B, and we would check the functional dependency `A -> B`

| A | B |
| -- | -- |
| 1 | X |
| 1 | X |
| 1 | Y |
| 2 | X | 

`unqiues` would result in:

| A | B |
| -- | -- |
| 1 | X |
| 1 | Y |
| 2 | X | 

`violations_stmt` would result in

| **A** | count(*) |
| -- | -- |
| 1 | 2 |
| 2 | 1 | 

Now we know that the key 1 is assigned to two different values and thus violates the FD.
However, now we need the shortcut to find all the values to which the violating keys are assigned so that we can output the nice error messages. This join results in 

| A | B |
| -- | -- |
| 1 | X |
| 1 | Y |

Does this explain the requirement for the join?

Reply:
Thanks a lot!